File: docs/examples/cpp/src/data_management_delete_records.cc
#include <reduct/client.h>
#include <cassert>
#include <chrono>

using reduct::IBucket;
using reduct::IClient;
using reduct::Error;
using std::chrono_literals::operator ""s;

int main() {
    // Create a client instance, then get or create a bucket
    auto client = IClient::Build("http://127.0.0.1:8383", {.api_token="my-token"});
    auto [bucket, create_err] = client->GetOrCreateBucket("my-bucket");
    assert(create_err == Error::kOk);

    //  Send some records to the "cpp-example"
    IBucket::Time ts = IBucket::Time::clock::now();
    auto err = bucket->Write("cpp-example", ts, [](auto rec) {
        rec->WriteAll("Some binary data");
    });
    assert(err == Error::kOk);

    err = bucket->Write("cpp-example", ts + 1s, [](auto rec) {
        rec->WriteAll("Some binary data");
    });
    assert(err == Error::kOk);

    // Delete a sole record
    err = bucket->RemoveRecord("cpp-example", ts);
    assert(err == Error::kOk);

    // Delete a batch of records
    auto [record_errors, http_err] = bucket->RemoveBatch("cpp-example", [ts](auto batch) {
        batch->AddRecord(ts);   // Already deleted, so this error will be in the errors map
        batch->AddRecord(ts + 1s);
    });

    assert(http_err == Error::kOk);
    assert(record_errors.begin()->second.code == 404);
    return 0;
}

File: docs/examples/py/src/data_management_delete_records.py
from time import time

from reduct import Client, Bucket, Batch


async def main():
    # Create a client with the base URL and API token
    async with Client("http://localhost:8383", api_token="my-token") as client:
        bucket: Bucket = await client.create_bucket("my-bucket", exist_ok=True)

        # Send some records to the "py-example"
        ts = time()
        await bucket.write(
            "py-example",
            b"Some binary data",
            ts
        )
        await bucket.write(
            "py-example",
            b"Some binary data",
            ts + 1
        )

        # Delete a sole record
        await bucket.remove_record("py-example", ts)

        # Delete a batch of records
        batch = Batch()
        batch.add(ts)  # already deleted, so this error will be in the errors map
        batch.add(ts + 1)

        errors = await bucket.remove_batch("py-example", batch)
        assert len(errors) == 1
        print(errors)

if __name__ == "__main__":
    import asyncio

    asyncio.run(main())
